SQL compatible Null Handling Part - Expressions and Storage Changes

nishantmonu51 opened the pull request:
Part 1 of Null Handling Changes -  https://github.com/druid-io/druid/pull/4754 which contains changes in Expressions and Storage Layer.

This PR contains changes to improve handling of null values in druid by treating nulls as missing values which don’t actually exist. This will make it more compatible to SQL standard and help with integration with other existing BI systems which support ODBC/JDBC. (Detailed description in proposal - https://github.com/druid-io/druid/issues/4349)

Changes include - 

## New Configuration 

* New configuration - **_"druid.generic.useDefaultValueForNull"_** in class NullValueHandlingConfig default to true for preserving old behavior.
* NullHandling - this a helper class with static methods for handling optional nullToEmpty / emptyToNull conversions for switching between old/new behavior. Without this static helper, we would need to inject NullValueHandlingConfig almost everywhere throughout the querying and indexing layer. 

## Storage Layer - 

* GenericIndexed - Added ability to distinguish between null/empty bytes. the value buffer for GenericIndexed contains element size followed by  the bytes, for null values size is written as -1 and for empty bytes size is 0. 
* String Columns - the dictionary is stores as a genericIndexed, which now can contain null value. An inverted index is created with the id assigned to null value. Null and empty strings will get different id's in dictionary encoding.
* Numeric Columns - The column is serialized as primitive objects of long/float/double type, a nullRowsBitmap is also stored which is used to distinguish rows having 0 and null values.  
* Complex columns - Complex columns use genericIndexed internally for storage, so supported nulls in genericIndexed make this work. the ObjectStrategy can choose to return null objects when deserializing data. 

## Indexing Layer - 

* Row - instead of returning primitive types return Float/Long/Double objects so that indexer can distinguish between null and 0 values.
* StringIndexer,DoubleIndexer,LongIndexer,FloatIndexer changed to distinguish between null and empty values. 
* ColumnValueSelector - added new method isNull() to check whether a value is null or not while aggregating. this is mainly done in order to avoid boxing/unboxing from primitive values while doing segment scans. 

## Expressions Null Handling
* For general calculations like sum, full expression is be considered as null if any of the components is null. 
* StringLiteral now supports null and and expressions containing non-quoted null is parsed as StringLiteral with null 
* Specifying a default value for null is supported by the use of NVL or IF clause to assign default values at query time.

## Backwards Compatibility 
* **_"druid.generic.useDefaultValueForNull"_** is added as a new config. The default value is set to true for backwards compatibility. 
*  code places in query/ingestion where we converted emptyToNull or nullToEmpty now use NullHandlingHelper which only does the conversion if backwards compatibility is enabled. 
* Storage layer changes are backwards compatible for GenericIndexed/String columns. 
* For Numeric columns new V2 serde implementations are introduced which also supports serde of nullRowsBitmaps. Previous implementations are kept unchanged to read any existing segments. 
* Note:- New serde are only used when Null Handling is enabled to enable rollback to a previous version when needed.

nishantmonu51 commented:
@leventov broken the PR into Expressions + Storage Changes. 
Follow up PR plan - 
1) Nullable Aggregator 
2) Null Handling in filters 
3) Null Handling in SQL Layer 
4) Enabling tests for null handling in CI and add api's like Strings.nullToEmpty and Strings.emptytoNull to forbidden api list. 

I handed all your comments in the previous big PR and replied to them individually. Please comment here in case i missed any. 

+cc @b-slim @gianm Please review.

nishantmonu51 commented:
@gianm: Also updated the PR to only use new serde for storage only when null handling is enabled. This would allow users to roll back to a previous release when null handling is not enabled.

leventov commented on common/src/main/java/io/druid/math/expr/Expr.java:
```diff
@@ -362,6 +362,13 @@ public ExprEval eval(ObjectBinding bindings)
+
+    // Result of any Binary expressions is null if any of the argument is null.
+    // e.g "select null * 2 as c;" or "select null + 1 as c;" will return null as per Standard SQL spec.
+    if (!NullHandling.useDefaultValuesForNull() && (leftVal.isNull() || rightVal.isNull())) {
```
I find it hard to mentally follow the phrase `useDefaultValuesForNull()`, plus negation in the beginning. What about making two methods with more vivid names instead,
 - `NullHandling.replace()`, the same as current `useDefaultValuesForNull()`,
 - `NullHandling.sqlCompatible()`, the same as current `!useDefaultValuesForNull()`?

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/BufferAggregator.java:
```diff
@@ -178,4 +178,23 @@ default void relocate(int oldPosition, int newPosition, ByteBuffer oldBuffer, By
+   * @param position offset within the byte buffer at which the aggregate value is stored
+   *
+   * @return true if the aggrgeated value is null otherwise false.
+   * For backwards compatibility, isNull() may return false even if {@link BufferAggregator#get(ByteBuffer, int)} returns null. Users of this method should account for this case.
```
This sentence is now not compatible with the new contract and the beginning of this javadoc comment.

nishantmonu51 replied:
fixed.

leventov commented on processing/src/main/java/io/druid/query/aggregation/ObjectAggregateCombiner.java:
```diff
@@ -26,4 +26,9 @@
 public abstract class ObjectAggregateCombiner<T> implements AggregateCombiner<T>, ObjectColumnSelector<T>
 {
+  @Override
+  public boolean isNull()
```
Makes sense to make the method `final` because object selectors must always return false.

leventov commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -327,18 +328,18 @@ public static Long getExactLongFromDecimalString(String decimalStr)
+  public static Number nullToZeroDouble(@Nullable Number number)
   {
-    return number == null ? ZERO_DOUBLE : number;
+    return number == null ? NullHandling.ZERO_DOUBLE : number;
```
Why changing references? There is `ZERO_DOUBLE` in this class. Same below

leventov commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -327,18 +328,18 @@ public static Long getExactLongFromDecimalString(String decimalStr)
   }
 
-  public static Double nullToZero(@Nullable Double number)
+  public static Number nullToZeroDouble(@Nullable Number number)
```
Why changing types and method name? Same below

nishantmonu51 replied:
reverted.

leventov commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -228,7 +229,7 @@ public boolean matches()
 
                 for (String dimensionValue : dimensionValues) {
-                  if (Objects.equals(Strings.emptyToNull(dimensionValue), value)) {
+                  if (Objects.equals(NullHandling.emptyToNullIfNeeded(dimensionValue), value)) {
```
Group By experts - @gianm / @jon-wei / @jihoonson - could you please check that this change is valid (same below)

leventov replied:
Also, a more general question -- how do you distinguish between empty strings that come from "old" segment format (and then should be converted to nulls), and empty strings that come from "new" segment format, and then empty string is a "proper" empty string, and shouldn't be converted?

Shouldn't use of `emptyToNullIfNeeded()` be localized *only* to `GenericIndexed` that handles old segment format?

nishantmonu51 replied:
I agree on the behavior of old data that empty strings should be converted to null when the segment is read, but that is not the case at present since full query layer is not capable of handling null values and at some places like filters we specifically match for empty strings. 
Note this PR only has partial changes to support new storage format and not touches the older format and query layer to make it short.

jon-wei replied:
I think the `NullHandling.emptyToNullIfNeeded` calls may be unnecessary here, since I believe this value matcher would only be used when filtering the result of a groupBy subquery, so the subquery results should have already applied the necessary conversions when running in the older "empty == null" mode

nishantmonu51 replied:
tried removing emptyToNull call, but that lead to Unit test failures for filters, in BaseFilterTest at selectColumnValuesMatchingFilterUsingRowBasedColumnSelectorFactory which specifically tries to test this, So it seems we may need this for the case tested there. 
since this PR does not introduce this as a new change in RowBasedColumnSelectorFactory, it may be ok to keep the check as it is.

leventov commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -410,19 +418,25 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
         public double getDouble()
         {
-          return row.get().getMetric(columnName).doubleValue();
+          return DimensionHandlerUtils.nullToZeroDouble(row.get().getMetric(columnName)).doubleValue();
```
Shouldn't this method throw exception, if the value is null, and SQL-compatible handling is on? (same below)

leventov commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -410,19 +418,25 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+        }
+
+        @Override
+        public boolean isNull()
```
Please move this method either before or after all getXxx() methods

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -410,19 +418,25 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
         public float getFloat()
         {
-          return row.get().getMetric(columnName).floatValue();
+          return DimensionHandlerUtils.nullToZeroDouble(row.get().getMetric(columnName)).floatValue();
```
`nullToZeroFloat()` should be used

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -410,19 +418,25 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
         public long getLong()
         {
-          return row.get().getMetric(columnName).longValue();
+          return DimensionHandlerUtils.nullToZeroDouble(row.get().getMetric(columnName)).longValue();
```
`nullToZeroLong()` should be used

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/BaseNullableColumnValueSelector.java:
```diff
@@ -0,0 +1,28 @@
+import io.druid.guice.annotations.PublicApi;
+
+@PublicApi
+public interface BaseNullableColumnValueSelector
```
Add javadoc to this class.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/BaseNullableColumnValueSelector.java:
```diff
@@ -0,0 +1,28 @@
+@PublicApi
+public interface BaseNullableColumnValueSelector
+{
+  boolean isNull();
```
Add javadoc to this method.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/ColumnSelectorBitmapIndexSelector.java:
```diff
@@ -180,7 +180,7 @@ public int getIndex(String value)
           // Return -2 for non-null values to match what the BitmapIndex implementation in BitmapIndexColumnPartSupplier
```
Please make this comment a javadoc comment of the enclosing method and make all class and method refs javadoc refs.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/ConstantColumnValueSelector.java:
```diff
@@ -86,4 +86,10 @@ public void inspectRuntimeShape(final RuntimeShapeInspector inspector)
+  @Override
+  public boolean isNull()
+  {
+    return objectValue == null;
```
Probably it should just return `false`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DimensionSelector.java:
```diff
@@ -150,12 +150,24 @@ default long getLong()
+  }
+
+  /**
+   * @deprecated always throws {@link UnsupportedOperationException}
```
This statement is false

nishantmonu51 replied:
removed.

leventov commented on processing/src/main/java/io/druid/segment/DimensionSelector.java:
```diff
@@ -150,12 +150,24 @@ default long getLong()
+  @Override
+  default boolean isNull()
+  {
+    throw new UnsupportedOperationException("DimensionSelector cannot be operated as numeric ColumnValueSelector" + this.getClass());
```
`DimensionSelector` is closer to be an "object" selector, then numeric. So I think this method should just return false. Indeed, it couldn't return `null` from getLong/Double/Float, but it *could* return null from it's getObject().

nishantmonu51 replied:
done.

nishantmonu51 commented on common/src/main/java/io/druid/common/config/NullHandling.java:
```diff
@@ -0,0 +1,106 @@
+
+  public static boolean isNullOrEquivalent(@Nullable String value)
+  {
+    return INSTANCE.isUseDefaultValuesForNull() ? Strings.isNullOrEmpty(value) : value == null;
```
It is still used at places like filters and aggregators. will add that in follow up PR. 
Have mentioned it in the follow up PR plan.

leventov replied:
May use replaceWithDefault() as in other methods in this class

nishantmonu51 replied:
done.

nishantmonu51 commented:
@leventov: have worked through your review comments, please check.

leventov commented on processing/src/main/java/io/druid/query/aggregation/BufferAggregator.java:
```diff
@@ -178,4 +178,24 @@ default void relocate(int oldPosition, int newPosition, ByteBuffer oldBuffer, By
+   * @param buf      byte buffer storing the byte array representation of the aggregate
+   * @param position offset within the byte buffer at which the aggregate value is stored
+   *
+   * @return true if the aggrgeated value is null otherwise false.
```
Typo in word `aggrgeated`, plus I suggest it to just follow the first sentence of this doc.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/query/aggregation/BufferAggregator.java:
```diff
@@ -178,4 +178,24 @@ default void relocate(int oldPosition, int newPosition, ByteBuffer oldBuffer, By
+   * @param position offset within the byte buffer at which the aggregate value is stored
+   *
+   * @return true if the aggrgeated value is null otherwise false.
+   * The default implementation always return false to enable smooth backward compatibility,
```
I think don't need to repeat this once again.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/BaseNullableColumnValueSelector.java:
```diff
@@ -0,0 +1,38 @@
+   * returns true if selected primitive value is null for {@link BaseFloatColumnValueSelector},
+   * {@link BaseLongColumnValueSelector} and {@link BaseDoubleColumnValueSelector} otherwise false.
+   */
+  boolean isNull();
```
Add `@CalledFromHotLoop`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DimensionSelector.java:
```diff
@@ -149,13 +149,21 @@ default long getLong()
    * This method is not the default implementation of {@link #getObject()} to minimize the chance that implementations
    * "forget" to override it with more optimized version.
    */
+  @Deprecated
```
Why?

nishantmonu51 replied:
removed. added by mistake while merging from another branch.

leventov commented on processing/src/main/java/io/druid/segment/DoubleColumnSelector.java:
```diff
@@ -61,6 +61,9 @@ default long getLong()
   @Override
   default Double getObject()
```
Annotate Nullable

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DoubleColumnSelector.java:
```diff
@@ -61,6 +61,9 @@ default long getLong()
   @Override
   default Double getObject()
   {
+    if (isNull()) {
```
Make similar changes in FloatColumnSelector and LongColumnSelector

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DoubleColumnSerializerV2.java:
```diff
@@ -0,0 +1,134 @@
+import java.nio.ByteOrder;
+import java.nio.channels.WritableByteChannel;
+
+public class DoubleColumnSerializerV2 implements GenericColumnSerializer
```
Add javadoc

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DoubleColumnSerializerV2.java:
```diff
@@ -0,0 +1,134 @@
+  @Override
+  public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    channel.write(ByteBuffer.wrap(Ints.toByteArray((int) writer.getSerializedSize())));
```
Use SerializerUtils.writeInt()

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DoubleDimensionIndexer.java:
```diff
@@ -21,17 +21,22 @@
 
 public class DoubleDimensionIndexer implements DimensionIndexer<Double, Double, Double>
 {
+  public static final Comparator DOUBLE_COMPARATOR = Comparators.<Double>naturalNullsFirst();
```
Add generic arg `<Double>`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/DoubleDimensionIndexer.java:
```diff
@@ -91,12 +96,23 @@ public DimensionSelector makeDimensionSelector(
 
-        if (dimIndex >= dims.length) {
+        if (dimIndex >= dims.length || dims[dimIndex] == null) {
           return 0.0;
```
throw something if sqlCompatible

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/FloatColumnSerializerV2.java:
```diff
@@ -0,0 +1,128 @@
+import java.nio.ByteOrder;
+import java.nio.channels.WritableByteChannel;
+
+public class FloatColumnSerializerV2 implements GenericColumnSerializer
```
Add javadoc

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/FloatColumnSerializerV2.java:
```diff
@@ -0,0 +1,128 @@
+  {
+    if (obj == null) {
+      nullRowsBitmap.add(rowCount);
+      writer.add(0L);
```
0f

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/FloatColumnSerializerV2.java:
```diff
@@ -0,0 +1,128 @@
+  @Override
+  public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    channel.write(ByteBuffer.wrap(Ints.toByteArray((int) writer.getSerializedSize())));
```
Use SerializerUtils

nishantmonu51 replied:
done

leventov commented on processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java:
```diff
@@ -21,18 +21,23 @@
 public class FloatDimensionIndexer implements DimensionIndexer<Float, Float, Float>
 {
 
+  public static final Comparator FLOAT_COMPARATOR = Comparators.<Float>naturalNullsFirst();
```
`<Float>`

nishantmonu51 replied:
done

leventov commented on processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java:
```diff
@@ -97,13 +102,20 @@ public float getFloat()
         return (Float) dims[dimIndex];
       }
 
+      @Override
```
Align this class with Double version. E. g. in Double version methods go in different order, and the logic is different.

nishantmonu51 replied:
done

leventov commented on processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java:
```diff
@@ -97,13 +102,20 @@ public float getFloat()
 
-        if (dimIndex >= dims.length) {
+        if (dimIndex >= dims.length || dims[dimIndex] == null) {
           return 0.0f;
```
throw something if sqlCompatible

nishantmonu51 replied:
done

leventov commented on processing/src/main/java/io/druid/segment/FloatDimensionMergerV9.java:
```diff
@@ -55,7 +55,12 @@ public FloatDimensionMergerV9(
   {
     final CompressionStrategy metCompression = indexSpec.getMetricCompression();
-    this.serializer = FloatColumnSerializer.create(segmentWriteOutMedium, dimensionName, metCompression);
+    this.serializer = IndexMergerV9.createFloatColumnSerializer(
```
Align with double version (there is a comment)

nishantmonu51 replied:
done.

nishantmonu51 commented:
@leventov: handled comments

jon-wei commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -407,22 +415,40 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
-          return row.get().getMetric(columnName).doubleValue();
+          Number metric = row.get().getMetric(columnName);
+          if (NullHandling.sqlCompatible() && metric == null) {
+            throw new IllegalStateException("Cannot return double for Null Value");
```
In follow-on patches, this exception would be avoided by having aggregators check "isNull()" before attempting to aggregate, is my understanding correct?

nishantmonu51 replied:
yes.

jon-wei commented on processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java:
```diff
@@ -407,22 +415,40 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
-          return row.get().getMetric(columnName).doubleValue();
+          Number metric = row.get().getMetric(columnName);
+          if (NullHandling.sqlCompatible() && metric == null) {
+            throw new IllegalStateException("Cannot return double for Null Value");
```
suggestion: rephrase the message and others like it as "Cannot return null value as double", I think that's clearer

nishantmonu51 replied:
sure, will do

nishantmonu51 replied:
done.

jon-wei commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -98,18 +98,20 @@
     {
-      return StringUtils.fromUtf8(buffer, numBytes);
+      if (numBytes < 0) {
+        // nulBytes will be -1 for null values.
```
Can you add this info to the format specification at the top of GenericIndexed as well?

Maybe use a constant for the -1 here with a name that indicates it's the marker for a null value

nishantmonu51 replied:
done.

jon-wei commented on processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java:
```diff
@@ -341,4 +341,9 @@ public static Float nullToZero(@Nullable Float number)
+
+  public static Number nullToZero(@Nullable Number number)
+  {
+    return number == null ? ZERO_DOUBLE : number;
```
In this proposal https://github.com/druid-io/druid/issues/4349 I saw you planned to add this feature:

> Users will be able to specify default values to replace NULL values at ingestion time.

Do you plan on adding that in a later patch? I'm wondering since this PR assumes zero will always be the default value.

nishantmonu51 replied:
yeah, that can be added easily as a separate patch/feature, first aiming at adding proper support for Nulls.

jon-wei commented:
Done reviewing for now, didn't have additional comments past GenericIndexed

nishantmonu51 commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -527,13 +572,13 @@ private LongColumnSerializer setupTimeWriter(SegmentWriteOutMedium segmentWriteO
       switch (type) {
         case LONG:
-          writer = LongColumnSerializer.create(segmentWriteOutMedium, metric, metCompression, longEncoding);
+          writer = createLongColumnSerializer(segmentWriteOutMedium, metric, metCompression, longEncoding, indexSpec);
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -98,18 +98,20 @@
     {
-      return StringUtils.fromUtf8(buffer, numBytes);
+      if (numBytes < 0) {
+        // nulBytes will be -1 for null values.
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/StringDimensionHandler.java:
```diff
@@ -77,6 +77,20 @@ public int compareSortedEncodedKeyComponents(int[] lhs, int[] rhs)
     return retVal;
   }
 
+  private boolean isNUllRow(int[] row, Indexed<String> encodings)
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/LongDimensionMergerV9.java:
```diff
@@ -58,7 +58,14 @@ protected void setupEncodedValueWriter(SegmentWriteOutMedium segmentWriteOutMedi
     final CompressionStrategy metCompression = indexSpec.getMetricCompression();
     final CompressionFactory.LongEncodingStrategy longEncoding = indexSpec.getLongEncoding();
-    this.serializer = LongColumnSerializer.create(segmentWriteOutMedium, dimensionName, metCompression, longEncoding);
+    // If using default values for null use LongColumnSerializer to allow rollback to previous versions.
```
dropped.

nishantmonu51 commented:
@jon-wei handled comments. 
@leventov: do you have any more comments ?

nishantmonu51 commented:
@leventov : do you have any more comments ? Can we merge this if there are no additional comments ?

leventov commented:
@nishantmonu51 I was just going to continue review. Please don't repeat your comments every few days, I see them. If I don't answer it means that I'm reviewing or going to review, but don't want to leave noisy comments like the one that you are reading at the moment.

leventov commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -194,19 +222,22 @@ public StringDimensionIndexer(MultiValueHandling multiValueHandling)
     if (dimValues == null) {
-      dimLookup.add(null);
-      encodedDimensionValues = null;
+      if (!dimLookup.contains(null)) {
```
Suggested to extract nullId variable and localize if-else (or use ternary)

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -194,19 +222,22 @@ public StringDimensionIndexer(MultiValueHandling multiValueHandling)
         encodedDimensionValues = EMPTY_INT_ARRAY;
       } else if (dimValuesList.size() == 1) {
-        encodedDimensionValues = new int[]{dimLookup.add(STRING_TRANSFORMER.apply(dimValuesList.get(0)))};
+        encodedDimensionValues = new int[]{dimLookup.add(EMPTY_TO_NULL_IF_NEEDED.apply(dimValuesList.get(0)))};
```
`EMPTY_TO_NULL_IF_NEEDED` is a function that is held in a static final constant and always called via .apply(). So it could be a static method.

nishantmonu51 replied:
done. changed to static method.

leventov commented on processing/src/main/java/io/druid/segment/column/DoublesColumn.java:
```diff
@@ -28,10 +29,12 @@
 public class DoublesColumn implements GenericColumn
```
I suggest to have two separate classes, `DoublesColumnWithNulls` and `DoublesColumnWithoutNulls` (that's a rename of this class. Or just keep it's method. Or choose other.) And in non-null version `isNull()` implementation is `return false;`. Need the best optimization for the common case.

nishantmonu51 replied:
added a check for emptyImmutableBitmap, isNull now also checks if the bitmap is empty or not. 
Also, I think we should optimize bitmap.get instead when the bitmap is empty to always return false, in case that is already not optimized, instead of making different classes at many places just for this optimization.

leventov replied:
It won't allow JVM to resolve this method call result statically and throw away branches in the upstream calling code. So it's very important to make the body of method *just* `return false;`.

One class could be extend another, overriding just isNull() method, so not so much code is copied and added.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/FloatsColumn.java:
```diff
@@ -29,10 +30,15 @@
 public class FloatsColumn implements GenericColumn
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/LongsColumn.java:
```diff
@@ -29,10 +30,15 @@
 public class LongsColumn implements GenericColumn
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java:
```diff
@@ -177,6 +176,13 @@ public ValueMatcher makeValueMatcher(Predicate<String> predicate)
+        @Override
+        public boolean isNull()
+        {
+          IndexedInts row = getRow();
```
Should be `return false`, DimensionSelector is analog of Object in isNull() semantics. isNull() is only relevant for primitive. Might be worth adding this to javadocs.

nishantmonu51 replied:
removed isNull, it was not needed to override it here.

leventov commented on processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java:
```diff
@@ -282,6 +288,12 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+        @Override
+        public boolean isNull()
+        {
+          return getObject() == null;
```
return false

nishantmonu51 replied:
done.

leventov replied:
omission

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarDoubles.java:
```diff
@@ -38,7 +39,7 @@
   void close();
 
-  default ColumnValueSelector<Double> makeColumnValueSelector(ReadableOffset offset)
+  default ColumnValueSelector<Double> makeColumnValueSelector(ReadableOffset offset, ImmutableBitmap nullValueBitmap)
```
It should check if the provided bitmap is empty, and return optimized impl then with `isNull() { return false; }`

nishantmonu51 replied:
added a check for empty bitmap.

leventov replied:
Same as in https://github.com/druid-io/druid/pull/5278#discussion_r166073102

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarFloats.java:
```diff
@@ -39,7 +40,7 @@
   void close();
 
-  default ColumnValueSelector<Float> makeColumnValueSelector(ReadableOffset offset)
+  default ColumnValueSelector<Float> makeColumnValueSelector(ReadableOffset offset, ImmutableBitmap nullValueBitmap)
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarLongs.java:
```diff
@@ -39,7 +40,7 @@
   void close();
 
-  default ColumnValueSelector<Long> makeColumnValueSelector(ReadableOffset offset)
+  default ColumnValueSelector<Long> makeColumnValueSelector(ReadableOffset offset, ImmutableBitmap nullValueBitmap)
```
Same

nishantmonu51 replied:
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/ObjectColumnSelector.java:
```diff
@@ -76,4 +76,17 @@ default long getLong()
+   */
+  @Deprecated
+  @Override
+  default boolean isNull()
```
We can do this in a follow up PR, created - https://github.com/druid-io/druid/issues/5346

nishantmonu51 replied:
done. changed ObjectColumnSelector, Don't see any benefit to refactor Long/Float/Double yet. Can refactor them later if required.

leventov commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -389,10 +383,62 @@ private void makeMetricsColumns(
+    }
+  }
+
+  static ColumnPartSerde createFloatColumnPartSerde(GenericColumnSerializer serializer, IndexSpec indexSpec)
```
#5347

nishantmonu51 commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -462,14 +457,7 @@ public int compare(Pair<Integer, PeekingIterator<String>> lhs, Pair<Integer, Pee
-                    return Strings.nullToEmpty(input);
-                  }
-                }
+                input -> NullHandling.nullToEmptyIfNeeded(input)
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/ObjectColumnSelector.java:
```diff
@@ -76,4 +76,17 @@ default long getLong()
   }
+
+  /**
+   * @deprecated This method is marked as deprecated in ObjectColumnSelector to minimize the probability of accidential
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/ObjectColumnSelector.java:
```diff
@@ -76,4 +76,17 @@ default long getLong()
+  @Override
+  default boolean isNull()
+  {
+    T value = getObject();
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/StringDimensionHandler.java:
```diff
@@ -77,6 +77,20 @@ public int compareSortedEncodedKeyComponents(int[] lhs, int[] rhs)
     return retVal;
   }
 
+  private boolean isNullRow(int[] row, Indexed<String> encodings)
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -58,13 +58,18 @@
 public class StringDimensionIndexer implements DimensionIndexer<Integer, int[], String>
 {
-  private static final Function<Object, String> STRING_TRANSFORMER = o -> o != null ? o.toString() : null;
+  private static final Function<Object, String> EMPTY_TO_NULL_IF_NEEDED = o -> o != null
```
changed to a method here. Since this is only used here, don;t see any value addition in moving it out to NulHandling, would like to keep it here only.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -58,13 +58,18 @@
+                                                                          : null;
+
+  private static final int ABSENT_VALUE_ID = -1;
   private static final int[] EMPTY_INT_ARRAY = new int[]{};
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -80,37 +85,59 @@ public DimensionDictionary()
     {
       synchronized (lock) {
-        return valueToId.size();
+        return idToValue.size();
```
done.

nishantmonu51 commented:
@leventov: replied to your comments.

leventov commented on processing/src/main/java/io/druid/query/aggregation/BufferAggregator.java:
```diff
@@ -178,4 +178,23 @@ default void relocate(int oldPosition, int newPosition, ByteBuffer oldBuffer, By
+   *
+   * @return true if the aggregated value is primitive long/double/float and aggregated value is null otherwise false.
+   */
+  @CalledFromHotLoop
```
This method shouldn't be annotated with `@CalledFromHotLoop`.

nishantmonu51 replied:
removed.

leventov commented on processing/src/main/java/io/druid/segment/NilColumnValueSelector.java:
```diff
@@ -85,6 +85,12 @@ public Class classOfObject()
   }
 
+  @Override
+  public boolean isNull()
```
getLong/Double/Float methods in NilColumnSelector should throw IllegalStateException, if `NullHandling.sqlCompatible()` is true. I suggest to
 1. Remove `final` from this class (it still couldn't be extended, because the constructor is private)
 2. Add private static subclass "SqlCompatible" that throws exceptions
 3. Return the instance of the "main" NilColumnSelector or "SqlCompatible" from instance() method, depending on `NullHandling.sqlCompatible()`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/LongDimensionIndexer.java:
```diff
@@ -94,13 +99,27 @@ public DimensionSelector makeDimensionSelector(
+      @Override
+      public boolean isNull()
+      {
+        if (NullHandling.replaceWithDefault()) {
```
Let's define clearer, should isNull() return depend on the current NullHandling configuration, or not. Because in some other places, at least in NilColumnSelector, it doesn't depend. Please update javadocs and review all impls.

nishantmonu51 replied:
isNull need not depend on NullHandling configuration, removed the if block.

leventov commented on processing/src/main/java/io/druid/segment/ObjectColumnSelector.java:
```diff
@@ -68,12 +68,23 @@ default double getDouble()
   {
     T value = getObject();
     if (value == null) {
       return 0;
```
I think it should throw exception, if NullHandling.sqlCompatible() is true, but if this selector is treated as e. g. long-typed selector, isNull() always returns false, the user doesn't expect getLong() to throw an exception then. Isn't there an incoherence in the API system that we built?

nishantmonu51 replied:
Well, the initial API was coherent in the sense that isNull implementation for ObjectColumnSelector used to return meaningful true/false depending on whether the object is null or not. It was changed based on your feedback to not allow users to call for optimizing the case when getObject() is heavy.

Since the ObjectColumnSelectors can be treated as object/long/double/float column, it would make sense to revert to initial implementation and return meaningful true/false from isNull for ObjectColumnSelectors. 

does this sound good to you ?

leventov replied:
I don't like both. It's important to get it right. I will return when I will review this PR until the end, ignore this comment for now.

leventov commented on processing/src/main/java/io/druid/segment/column/DoublesColumnWithNulls.java:
```diff
@@ -0,0 +1,52 @@
+  @Override
+  public boolean isNull(int rowNum)
+  {
+    return nullValueBitmap.get(rowNum);
```
`nullValueBitmap ` should be inspected in `inspectRuntimeShape()`

nishantmonu51 replied:
added.

leventov commented on processing/src/main/java/io/druid/segment/column/FloatsColumnWithNulls.java:
```diff
@@ -0,0 +1,52 @@
+  @Override
+  public boolean isNull(int rowNum)
+  {
+    return nullValueBitmap.get(rowNum);
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/LongsColumnWithNulls.java:
```diff
@@ -0,0 +1,52 @@
+  @Override
+  public boolean isNull(int rowNum)
+  {
+    return nullValueBitmap.get(rowNum);
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarDoubles.java:
```diff
@@ -54,6 +55,44 @@ public double getDouble(int offset)
+    return new HistoricalDoubleColumnSelector();
+  }
+
+  default ColumnValueSelector<Double> makeColumnValueSelector(ReadableOffset offset, ImmutableBitmap nullValueBitmap)
```
I think it's safer when there is just one method, and it checks if the bitmap is empty internally, and if it is, it returns optimized implementation. Otherwise it's easier to make a mistake in some other places, and also it creates two overloaded methods which differ by the presence of the last parameter, so you may want to pass it, but just forget, but the code will still compile and run. Same comments for long and float.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarDoubles.java:
```diff
@@ -54,6 +55,44 @@ public double getDouble(int offset)
+      @Override
+      public boolean isNull()
+      {
+        return nullValueBitmap.get(offset.getOffset());
```
Should be inspected. Same for long and float.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarFloats.java:
```diff
@@ -49,6 +50,45 @@ public float getFloat()
+
+  default ColumnValueSelector<Float> makeColumnValueSelector(ReadableOffset offset, ImmutableBitmap nullValueBitmap)
+  {
+    final boolean hasNulls = !nullValueBitmap.isEmpty();
```
Please align the logic with double. And if there is just one method, it checks anyway and returns an optimized impl.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/ColumnarLongs.java:
```diff
@@ -39,6 +40,39 @@
   @Override
   void close();
 
+  default ColumnValueSelector<Long> makeColumnValueSelector(ReadableOffset offset, ImmutableBitmap nullValueBitmap)
```
Align logic and method order with double and float.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -53,7 +53,7 @@
  * byte 1: version (0x1)
  * byte 2 == 0x1 =>; allowReverseLookup
- * bytes 3-6 =>; numBytesUsed
+ * bytes 3-6 =>; numBytesUsed, It will be -1 for null values.
```
This addition should be at line 59, not here

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -97,19 +99,23 @@
     }
 
     @Override
     public byte[] toBytes(String val)
```
This method should be annotated `@Nullable`

leventov replied:
Not addressed

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -97,19 +99,23 @@
     public String fromByteBuffer(final ByteBuffer buffer, final int numBytes)
     {
-      return StringUtils.fromUtf8(buffer, numBytes);
+      if (numBytes < 0) {
```
It's hard to comprehend the logic of this method. Let's clarify concerns. There are two possible formats: legacy and the new, nullable, and two possible configurations: sqlCompatible and replaceWithDefault. So 4 options. Could you please write what happens in all 4 of them for two cases: numBytes < 0 and numBytes = 0. (in a comment in this method)

nishantmonu51 replied:
added comment.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -496,11 +497,11 @@ public int indexOf(T value)
           allowReverseLookup = false;
         }
 
-        // for compatibility with the format, but this field is unused
```
I don't like this design. It will make harder to remove this (in fact unused) field in the future. Instead, I suggest to encode nullness somewhere by applying bit inverse (~) to some field and writing it so.

nishantmonu51 replied:
agreed, this field was unused but that does not mean we cannot use it to store nullability now when we need to do that. I don't see a need to remove this field in future since it is now being used.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -413,11 +414,11 @@ public int size()
-      final int size = endOffset - startOffset;
+      final int size = endOffset > startOffset
+                       ? endOffset - startOffset
+                       : copyValueBuffer.get(startOffset - Ints.BYTES);
```
Integer.BYTES

nishantmonu51 replied:
How is this going to make any difference ?

leventov replied:
Consistency

nishantmonu51 replied:
I am not sure which consistency are we talking about here. I think there is nothing wrong with using Integer.BYTES or Ints.BYTES, when both are essentially same and both are already used in druid at many places. 
If you look at the complete GenericIndexed class Ints.BYTES is used at lots of places, So i think it should be ok to use the same here. 

It would be great if we please refrain from adding comments that do not contribute in a meaningful way in improving the PR, but they are just there to line up with whatever personal coding preference one has. 

Another way to add such comments is to prefix them with "optional:..." so that the PR owner knows that he can choose to handle or ignore the comment without blocking his PR.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -563,7 +564,7 @@ public T get(final int index)
 
         if (index == 0) {
-          startOffset = 4;
+          startOffset = Ints.BYTES;
```
Integer.

leventov commented on processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java:
```diff
@@ -79,6 +79,9 @@ public BitmapFactory getBitmapFactory()
     @Override
     public ImmutableBitmap fromByteBuffer(ByteBuffer buffer, int numBytes)
```
Nullable

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/filter/Filters.java:
```diff
@@ -132,7 +133,7 @@ public static ValueMatcher makeValueMatcher(
         );
 
-    return selector.getColumnSelectorStrategy().makeValueMatcher(selector.getSelector(), value);
+    return selector.getColumnSelectorStrategy().makeValueMatcher(selector.getSelector(), NullHandling.emptyToNullIfNeeded(value));
```
This line is longer than 120 cols

nishantmonu51 replied:
formatted.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -97,19 +99,28 @@
+      //   1. numBytes will be -1 for null value, return value will be null
+      //   2. numBytes will be 0 for empty string, return value will be empty string
+      // For Legacy null handling when nulls and empty string are considered same -
+      //     numBytes will be 0 for both empty string and null and return value will be null for both.
```
If so, could you please add something like `if (nullHandling.replaceWithDefaults()) throw new IllegalStateException("not expected")` in the `if (numBytes < 0) {` branch?

leventov replied:
Also, I think it's better to keep -1/null logic completely out of `ObjectStrategy` impls (like this one), and keep passing only non-negative numBytes into fromByteBuffer() method.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -57,7 +57,7 @@
  * bytes 10-((numElements * 4) + 10): integers representing *end* offsets of byte serialized values
  * bytes ((numElements * 4) + 10)-(numBytesUsed + 2): 4-byte integer representing length of value, followed by bytes
- * for value
+ * for value. length of value will be -1 for null values.
```
Actually, better to say that this field has no meaning, if next offset is strictly greater than the current offset, and if they are the same, -1 at this field means `null`, and 0 at this field means some object (potentially non-null - e. g. in the string case), that is serialized as an empty sequence of bytes.

To align this text with reality.

nishantmonu51 replied:
removed this.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -413,11 +421,11 @@ public int size()
     T bufferedIndexedGet(ByteBuffer copyValueBuffer, int startOffset, int endOffset)
     {
-      final int size = endOffset - startOffset;
+      final int size = endOffset > startOffset
```
It passes -1 as numBytes down into ObjectStrategies. I suggest to not do so.
```java
int size = endOffset - startOffset;
if (size == 0 && copyValueBuffer.get(startOffset - Ints.BYTES) == NULL_VALUE_SIZE_MARKER) {
  return null;
}
```
etc.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/RoaringBitmapSerdeFactory.java:
```diff
@@ -77,8 +78,12 @@ public BitmapFactory getBitmapFactory()
+    @Nullable
     public ImmutableBitmap fromByteBuffer(ByteBuffer buffer, int numBytes)
     {
+      if (numBytes == 0) {
```
Probably shouldn't be added, `GenericIndexed` should handle nullability with -1's.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -151,22 +151,28 @@ public static ColumnSelectorFactory makeColumnSelectorFactory(
             public long getLong()
             {
-              return in.get().getMetric(column).longValue();
+              return DimensionHandlerUtils.nullToZero(in.get().getMetric(column)).longValue();
```
Should be similar to `RowBasedColumnSelectorFactory`.

nishantmonu51 replied:
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/NilColumnValueSelector.java:
```diff
@@ -85,9 +88,43 @@ public Class classOfObject()
+  @Override
+  public boolean isNull()
+  {
+    return false;
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -54,17 +54,25 @@
+  }
+
+  private static final int ABSENT_VALUE_ID = -1;
+  private static final int[] EMPTY_INT_ARRAY = IntArrays.EMPTY_ARRAY;
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/serde/DoubleGenericColumnSupplier.java:
```diff
@@ -20,23 +20,34 @@
   public GenericColumn get()
   {
-    return new DoublesColumn(column.get());
+    if (nullValueBitmap.isEmpty()) {
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/column/DoublesColumnWithNulls.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * DoublesColumn with null values.
+ */
+public class DoublesColumnWithNulls extends DoublesColumn
```
done. Added asserts to all selectors and expressions also where we were converting nulls to primitive default values.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarDoubles.java:
```diff
@@ -33,35 +34,74 @@
+        @Override
+        public double getDouble()
+        {
+          return ColumnarDoubles.this.get(offset.getOffset());
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarDoubles.java:
```diff
@@ -33,35 +34,74 @@
+        @Override
+        public double getDouble(int offset)
+        {
+          return ColumnarDoubles.this.get(offset);
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarFloats.java:
```diff
@@ -33,35 +34,75 @@
-      {
-        return ColumnarFloats.this.get(offset);
-      }
+        @Override
```
not sure how this helps, but still rearranged.

leventov replied:
When "specialization" classes (Long/Float/Double) are identical except the "Long/Float/Double" word, it's easier to review changes, to apply changes by analogy, and it doesn't raise any questions. When they are different, it's always a question if they are different because there is some specific logic of a specific type (sometimes there is), or for no actual reason.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarFloats.java:
```diff
@@ -33,35 +34,75 @@
+        @Override
+        public float getFloat()
+        {
+          return ColumnarFloats.this.get(offset.getOffset());
```
added

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarFloats.java:
```diff
@@ -33,35 +34,75 @@
+        @Override
+        public double getDouble(int offset)
+        {
+          return ColumnarFloats.this.get(offset);
```
added

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarFloats.java:
```diff
@@ -33,35 +34,75 @@
+        }
+
+        @Override
+        public boolean isNull()
```
added

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarLongs.java:
```diff
@@ -33,35 +34,75 @@
+        @Override
+        public long getLong()
+        {
+          return ColumnarLongs.this.get(offset.getOffset());
```
added

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/ColumnarLongs.java:
```diff
@@ -33,35 +34,75 @@
+        @Override
+        public double getDouble(int offset)
+        {
+          return ColumnarLongs.this.get(offset);
```
added

leventov commented on processing/src/main/java/io/druid/segment/DimensionSelector.java:
```diff
@@ -149,6 +149,13 @@ default long getLong()
     return 0L;
```
maybe need to add assertion, let's revisit https://github.com/druid-io/druid/issues/4888

leventov commented on processing/src/main/java/io/druid/segment/DoubleColumnSerializerV2.java:
```diff
@@ -0,0 +1,139 @@
+  @Override
+  public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    SerializerUtils.writeInt(channel, (int) writer.getSerializedSize());
```
Suggested `Ints.checkedCast()`

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/FloatColumnSerializerV2.java:
```diff
@@ -0,0 +1,139 @@
+  @Override
+  public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    SerializerUtils.writeInt(channel, (int) writer.getSerializedSize());
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/LongColumnSerializerV2.java:
```diff
@@ -0,0 +1,144 @@
+  @Override
+  public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
+  {
+    SerializerUtils.writeInt(channel, (int) writer.getSerializedSize());
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/NilColumnValueSelector.java:
```diff
@@ -28,9 +29,11 @@
  */
 @PublicApi
-public final class NilColumnValueSelector implements ColumnValueSelector
+public class NilColumnValueSelector implements ColumnValueSelector
```
Javadoc of getDouble/Long/Float is outdated.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/FloatsColumn.java:
```diff
@@ -19,18 +19,32 @@
+  final ColumnarFloats column;
 
-  public FloatsColumn(final ColumnarFloats column)
+  protected FloatsColumn(final ColumnarFloats column)
```
Please align with Double (no protected)

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/LongsColumn.java:
```diff
@@ -19,18 +19,32 @@
+  final ColumnarLongs column;
 
-  public LongsColumn(final ColumnarLongs column)
+  protected LongsColumn(final ColumnarLongs column)
```
Same

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/column/DoublesColumnWithNulls.java:
```diff
@@ -0,0 +1,80 @@
+  @Override
+  public float getFloatSingleValueRow(int rowNum)
+  {
+    assert !isNull(rowNum);
```
Why checks here don't check replaceWithDefaults()? Also, are they needed, or always duplicated, because `DoublesColumn` is always wrapped as some ColumnValueSelector, which also does it's own checks? Same questions for other methods and long/float classes.

nishantmonu51 replied:
doublesColumnwithNulls are only created when null handling is enabled and null values are present. so no need to check for replaceWithdefault here.
Agreed, these checks may be redundant, but since they are assertions to catch issues with writing code, i added them based on your previous comments.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -57,7 +56,9 @@
  * bytes 7-10 =>; numElements
  * bytes 10-((numElements * 4) + 10): integers representing *end* offsets of byte serialized values
  * bytes ((numElements * 4) + 10)-(numBytesUsed + 2): 4-byte integer representing length of value, followed by bytes
```
", followed by bytes for value" is not part of "This field has no meaning..."

nishantmonu51 replied:
changed.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -370,11 +369,10 @@ public void writeTo(WritableByteChannel channel, FileSmoosher smoosher) throws I
 
   private T copyBufferAndGet(ByteBuffer valueBuffer, int startOffset, int endOffset)
```
The logic of this method should be the same as `bufferedIndexedGet()`, but now it could pass -1 to `strategy.fromByteBuffer()`.

Please review this aspect again across `GenericIndexed` and `GenericIndexedWriter`.

nishantmonu51 replied:
done.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java:
```diff
@@ -214,9 +214,8 @@ public void write(T objectToWrite) throws IOException
-    // for compatibility with the format (see GenericIndexed javadoc for description of the format), but this field is
-    // unused.
-    valuesOut.writeInt(0);
+
```
Please keep the comment about compatibility and that currently this unused field is used as opportunistically to store nullness marker, but in a better format this info should take 1 bit.

nishantmonu51 replied:
added back.

leventov commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -98,18 +101,16 @@
     public String fromByteBuffer(final ByteBuffer buffer, final int numBytes)
     {
-      return StringUtils.fromUtf8(buffer, numBytes);
+      return NullHandling.emptyToNullIfNeeded(StringUtils.fromUtf8Nullable(buffer, numBytes));
```
Should use just `fromUtf8()`, as before

nishantmonu51 replied:
done.

nishantmonu51 commented on extensions-contrib/virtual-columns/src/main/java/io/druid/segment/MapVirtualColumn.java:
```diff
@@ -192,6 +192,12 @@ public long getLong()
       return 0L;
```
isNull for this class always returns false, IMO, this method should never be called at all and we should always throw unsupported operation exception irrespective of whether null handling is enabled or not. Anyways, adding assert that replaceWithDefault is enabled when these methods are called without changing any older behavior.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/ConstantColumnValueSelector.java:
```diff
@@ -86,4 +86,10 @@ public void inspectRuntimeShape(final RuntimeShapeInspector inspector)
+  @Override
+  public boolean isNull()
+  {
+    return false;
```
handled in ExpressionSelectors, we will use NilColumnSelector for null values. 
still always returning false here as the none primitive values in this class can be null.

nishantmonu51 commented on processing/src/main/java/io/druid/query/aggregation/BufferAggregator.java:
```diff
@@ -178,4 +178,22 @@ default void relocate(int oldPosition, int newPosition, ByteBuffer oldBuffer, By
+  /**
+   * returns true if aggregator's output type is primitive long/double/float and aggregated value is null,
+   * but when aggregated output type is Object, this method always returns false,
+   * and users are advised to check nullability for the object returned by {@link #get(ByteBuffer, int)} ()}
```
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -255,13 +262,15 @@ public final long asLong()
+        assert NullHandling.replaceWithDefault();
         return 0.0;
       }
 
```
done.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -243,6 +249,7 @@ public final ExprType type()
+        assert NullHandling.replaceWithDefault();
         return 0;
       }
 
```
added.

nishantmonu51 commented on common/src/main/java/io/druid/math/expr/ExprEval.java:
```diff
@@ -130,18 +133,21 @@ private NumericExprEval(Number value)
     @Override
     public final int asInt()
     {
+      assert NullHandling.replaceWithDefault() || !isNull();
```
removed.

nishantmonu51 commented:
@leventov: addressed your comments. please continue the review.

leventov commented on processing/src/main/java/io/druid/segment/serde/FloatGenericColumnPartSerdeV2.java:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ */
+public class FloatGenericColumnPartSerdeV2 implements ColumnPartSerde
```
Ensure that Double, Float and Long have the same method order (and code) and differ only by the type

nishantmonu51 replied:
changed method order to existing order in FloatGenericColumnPartSerde and 
also compared each other they are same. 
also, if you think same method ordering is really required and we need to enforce that, ould be great if we can add a checkstyle check for this. 

IMO, making the individual method order same in different classes is not something that need not to be enforced.

leventov replied:
It couldn't be enforced easily, also those specializations need to have some minor differences (apart from the type) sometimes.

nishantmonu51 replied:
If you really think having same method ordering across multiple classes is valuable and worth a developer investing his time in rearranging methods and comparing multiple classes, would be great if you can add this to contribution guidelines on druid docs. 

Would really help in decreasing back and forth on future PRs and decrease the amount of iterations we need to go through w.r.t method orderings, which IMO do not add any value to the feature being added.

nishantmonu51 replied:
Created - https://github.com/druid-io/druid/issues/5407

leventov commented on processing/src/main/java/io/druid/segment/serde/LongGenericColumnPartSerdeV2.java:
```diff
@@ -0,0 +1,159 @@
+
+/**
+ */
+public class LongGenericColumnPartSerdeV2 implements ColumnPartSerde
```
Double, Float and Long look very similar, could a common base be extracted to reduce the amount of boilerplate?

nishantmonu51 replied:
this is out of scope of this PR, we can do this in a separate PR.

leventov replied:
Please create an issue

nishantmonu51 replied:
https://github.com/druid-io/druid/issues/5405

leventov commented on processing/src/main/java/io/druid/segment/virtual/ExpressionColumnValueSelector.java:
```diff
@@ -75,4 +75,10 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  @Override
+  public boolean isNull()
+  {
+    return getObject() == null;
```
According to the current way how those methods are implemented, it should be `return false`

nishantmonu51 replied:
IMO this should be getObject.isNull() instead as the getObject for this class would never return null instead an ExprEval. changing to getObject.isNull()

leventov replied:
There is an incoherency. This is a selector of `ExprEval`, not of `ExprEval.value()`. This class is used only in `ExpressionSelectors.makeExprEvalSelector()`. This incoherency already causes problems in `AggregatorUtil`: `ExpressionSelectors.makeExprEvalSelector()` is used there only to call `getObject()`, not taking advantage of `ConstantColumnValueSelector` optimization.

To make things more coherent, either
 - `ExpressionSelectors.makeExprEvalSelector()` should return `ColumnValueSelector<T>`, not of `ExprEval` (with according changes)
 - Or `Supplier<ExprEval>` should be returned from this method, enforcing all clients to go via `get().asDouble()` path (and `ConstantColumnValueSelector` should adapt accordingly).

FYI @gianm

nishantmonu51 replied:
created https://github.com/druid-io/druid/issues/5406 for follow up. 
This need not block this PR.

leventov commented on processing/src/main/java/io/druid/segment/virtual/ExpressionColumnValueSelector.java:
```diff
@@ -75,4 +75,10 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
     inspector.visit("expression", expression);
     inspector.visit("bindings", bindings);
   }
+
```
getDouble/Float/Long need getObject().isNull() and replaceWithDefault() checks

nishantmonu51 replied:
expressions already have checks, no need to add redundant checks in expression selectors.

leventov replied:
Please add comments for readers, expecting asserts

nishantmonu51 replied:
added.

leventov commented on processing/src/main/java/io/druid/segment/virtual/ExpressionSelectors.java:
```diff
@@ -77,6 +78,12 @@ public double getDouble()
         return baseSelector.getDouble();
```
checks?

nishantmonu51 replied:
Expressions already have the checks, don't think we need to add redundant checks here.

leventov replied:
Please add comments for readers, expecting asserts

nishantmonu51 replied:
added.

leventov commented on processing/src/main/java/io/druid/segment/virtual/SingleLongInputCachingExpressionColumnValueSelector.java:
```diff
@@ -143,4 +143,10 @@ private ExprEval eval(final long value)
     bindings.set(value);
```
Checks in this class in getDouble/Long?

@nishantmonu51 please review once again ALL classes that implement `ColumnValueSelector` in the Druid project, without skipping those which extend each other, if they need those checks or not.

nishantmonu51 replied:
I intentionally skipped checks for expression selectors, as all subclasses of ExprEval already have checks.

leventov replied:
Please add comments for readers, expecting asserts

nishantmonu51 replied:
added.

leventov commented on processing/src/main/java/io/druid/segment/virtual/SingleStringInputCachingExpressionColumnValueSelector.java:
```diff
@@ -133,6 +133,12 @@ private ExprEval eval()
     return expression.eval(bindings);
   }
 
```
Checks in this class?

nishantmonu51 replied:
ExprEval already have checks.

leventov replied:
Please add comments for readers, expecting asserts

nishantmonu51 replied:
added.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -101,13 +102,13 @@
     public String fromByteBuffer(final ByteBuffer buffer, final int numBytes)
     {
-      return NullHandling.emptyToNullIfNeeded(StringUtils.fromUtf8Nullable(buffer, numBytes));
+      return StringUtils.fromUtf8(buffer, numBytes);
```
it is not needed here.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/serde/FloatGenericColumnPartSerde.java:
```diff
@@ -65,15 +65,15 @@ public static SerializerBuilder serializerBuilder()
   {
     private ByteOrder byteOrder = null;
-    private FloatColumnSerializer delegate = null;
+    private Serializer delegate = null;
```
IndexMergerV9.createFloatColumnPartSerde passes a GenericColumnSerializer instead of FloatColumnSerializer, this is done in order to not have duplicate method to create column part serdes in IndexMergerV9 and Float/Long/DoubleMergerV9 classed.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/serde/DoubleGenericColumnPartSerdeV2.java:
```diff
@@ -0,0 +1,162 @@
+  private final BitmapSerdeFactory bitmapSerdeFactory;
+
+  @JsonCreator
+  public static DoubleGenericColumnPartSerdeV2 getDoubleGenericColumnPartSerde(
```
changed the order to be same as existing class DoubleGenericColumnPartSerde.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/NilColumnValueSelector.java:
```diff
@@ -102,19 +102,27 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
   {
-
+    /**
+     * Always throws IllegalStateException.
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/serde/DoubleGenericColumnPartSerdeV2.java:
```diff
@@ -0,0 +1,162 @@
+      @Nullable @JsonProperty("bitmapSerdeFactory") BitmapSerdeFactory bitmapSerdeFactory
+  )
+  {
+    return new DoubleGenericColumnPartSerdeV2(byteOrder,
```
done.

nishantmonu51 commented on processing/src/main/java/io/druid/segment/serde/DoubleGenericColumnPartSerdeV2.java:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  @JsonProperty
+  public ByteOrder getByteOrder()
```
reordered.

nishantmonu51 commented:
@leventov: updated based on your comments, Hope you have completed the review and we can get this merged sooner.

leventov commented:
@nishantmonu51 please don't rebase the whole branch, it breaks previous review conversations and makes it harder to track changes, merge master into your feature branch instead.

leventov commented on processing/src/main/java/io/druid/segment/virtual/ExpressionSelectors.java:
```diff
@@ -77,6 +78,12 @@ public double getDouble()
       }
 
+      @Override
+      public boolean isNull()
```
Arrange this method after all getXxx().

nishantmonu51 replied:
Rearranged.

leventov commented on processing/src/main/java/io/druid/segment/serde/DoubleGenericColumnPartSerdeV2.java:
```diff
@@ -36,36 +36,27 @@
-                                              : new BitmapSerde.LegacyBitmapSerdeFactory(), null
+    return new DoubleGenericColumnPartSerdeV2(
+        byteOrder,
+        bitmapSerdeFactory != null
```
Code around here is not formatted properly (compare with `DoubleGenericColumnPartSerdeV2`)

nishantmonu51 replied:
done.

nishantmonu51 commented:
@leventov: Added comments to expression selectors and changed the formatting as mentioned. 
Also created follow up issues for your comments that need not block this PR. 

Could you also add a summarized comment regarding your thoughts on this PR ? 
Have you completed the review and were above comments were your final set of comments ? 
are there any major changes apart from the nitpicks that you think are needed before this can be merged ? 
Any thoughts on, How long do you think this PR is going to take to be ready to be merged in your opinion ?

leventov commented:
@nishantmonu51 could be merged after IntelliJ and Travis CI is made green. Could you please take stewardship of #5406?

I also want to note that `isNull()` return type and the behaviour of this method is not a done question, let's not use the fact that it's already committed as an argument for keeping the things as is.

nishantmonu51 commented:
Intellij and travis are green now.

dclim commented:
@nishantmonu51 for the release notes, can you confirm:

If `druid.generic.useDefaultValueForNull` is set to false, will the produced segments be incompatible with versions of Druid before 0.13.0?
